File: app/schemas/catalog.py
"""
Pydantic-схемы для каталога: категории и товары.
"""
from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel


class CategoryShort(BaseModel):
    """Краткая информация о категории (вложена в товар)."""
    id: UUID
    name: str
    slug: str

    class Config:
        from_attributes = True


class CategoryOut(BaseModel):
    """Категория с дочерними элементами для построения дерева."""
    id: UUID
    name: str
    slug: str
    image_url: Optional[str] = None
    children: Optional[list["CategoryOut"]] = None

    class Config:
        from_attributes = True


class ProductShort(BaseModel):
    """Краткое представление товара в списке."""
    id: UUID
    name: str
    price: float
    quantity: int
    image_urls: list = []
    category: Optional[CategoryShort] = None

    class Config:
        from_attributes = True


class ProductDetail(BaseModel):
    """Полная информация о товаре."""
    id: UUID
    name: str
    description: str = ""
    price: float
    quantity: int
    image_urls: list = []
    attributes: dict = {}
    category: Optional[CategoryShort] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ProductListResponse(BaseModel):
    """Ответ списка товаров с пагинацией."""
    items: list[ProductShort]
    total: int
    page: int
    limit: int

File: frontend/src/api/catalog.ts
import api from './client';

export interface CategoryNode {
  id: string;
  name: string;
  slug: string;
  image_url: string | null;
  children: CategoryNode[] | null;
}

export interface CategoryShort {
  id: string;
  name: string;
  slug: string;
}

export interface ProductShort {
  id: string;
  name: string;
  price: number;
  quantity: number;
  image_urls: string[];
  category: CategoryShort | null;
}

export interface ProductDetail {
  id: string;
  name: string;
  description: string;
  price: number;
  quantity: number;
  image_urls: string[];
  attributes: Record<string, string>;
  category: CategoryShort | null;
  created_at: string;
  updated_at: string;
}

export interface ProductListResponse {
  items: ProductShort[];
  total: number;
  page: number;
  limit: number;
}

export interface ProductFilters {
  search: string;
  category_id?: string;
  min_price?: number;
  max_price?: number;
  in_stock?: boolean;
  sort?: string;
  page?: number;
  limit?: number;
}

export const catalogApi = {
  getCategories(): Promise<{ data: CategoryNode[] }> {
    return api.get('/api/v1/categories');
  },

  getProduct(id: string): Promise<{ data: ProductDetail }> {
    return api.get(`/api/v1/products/${id}`);
  },

  getProducts(filters: ProductFilters): Promise<{ data: ProductListResponse }> {
    const params: Record<string, string | number | boolean> = {
      search: filters.search,
    };
    if (filters.category_id) params.category_id = filters.category_id;
    if (filters.min_price !== undefined) params.min_price = filters.min_price;
    if (filters.max_price !== undefined) params.max_price = filters.max_price;
    if (filters.in_stock !== undefined) params.in_stock = filters.in_stock;
    if (filters.sort) params.sort = filters.sort;
    if (filters.page) params.page = filters.page;
    if (filters.limit) params.limit = filters.limit;
    return api.get('/api/v1/products', { params });
  },
};